core/src/main/java/org/elasticsearch/transport/netty/MessageChannelHandler.java

```diff
@@ -342,4 +313,66 @@ public void onFailure(Throwable e) {
             }
         }
     }
+
+    /**
+     * Internal helper class to store characteristic offsets of a buffer during processing
+     */
+    protected static final class Marker {
+        private final ChannelBuffer buffer;
+        private final int size;
+        private final int markedReaderIndex;
+        private final int expectedReaderIndex;
+
+        public Marker(ChannelBuffer buffer) {
+            this.buffer = buffer;
+            this.size = buffer.getInt(buffer.readerIndex() - 4);
```

Comment:
this 4 is because we just read an int of the buffer in SizeHeaderFrameDecoder  (we should add a comment). I also prefer to have the logic in this class as two utility methods validateRequestIsRead / validateResponseIsRead which get all the info they need. To understand you need to see when the constructor is called anyway. If you feel strongly it should be class, then I'm fine but please doc it.